### PR TITLE
Non-pushable walking NPC

### PR DESCRIPTION
**Pull request by pzaj2**
This will allow marking NPCs as non-pushable even if they can walk on their own. "pushable" XML attribute is optional.

### CODE REVIEW

**Comment by jo3bingham** [comment]
Spacing.

**Comment by pzaj2** [comment]
Alright, you're right. My editor was showing them just fine. Fixed (I believe).

File: src/npc.cpp [hunk 1]
```diff
@@ -153,6 +154,12 @@ bool Npc::loadFromXml()
 		baseSpeed = 100;
 	}
 
+        if((attr = npcNode.attribute("pushable"))) {
+            pushable = attr.as_bool();
+        } else {
```
Comment (EvulMastah) [inline src/npc.cpp]: Not needed. It is set to true by default in `reset()`.

Also, space after `if` and correct spacing here (use tabs instead of spaces).

Comment (pzaj2) [inline src/npc.cpp]: here you go.

File: src/game.cpp [hunk 2]
```diff
@@ -738,7 +738,7 @@ void Game::playerMoveCreature(Player* player, Creature* movingCreature, const Po
 			}
 
 			Npc* movingNpc = movingCreature->getNpc();
-			if (movingNpc && !Spawns::isInZone(movingNpc->getMasterPos(), movingNpc->getMasterRadius(), toPos)) {
+			if (movingNpc && (!movingNpc->isPushable() || !Spawns::isInZone(movingNpc->getMasterPos(), movingNpc->getMasterRadius(), toPos))) {
```
Comment (ranisalt) [inline src/game.cpp]: You can use De Morgan and change to `!(movingNpc->isPushable() && Spawns::isInZone(...))` instead.

File: src/npc.h [hunk 3]
```diff
@@ -118,7 +118,7 @@ class Npc final : public Creature
 		}
 
 		bool isPushable() const final {
-			return walkTicks > 0;
+			return walkTicks > 0 && pushable;
```
Comment (ranisalt) [inline src/npc.h]: Test for pushability first, `pushable && walkTicks > 0`. Actually, since `walkTicks` is unsigned, you could test `!= 0` instead.

Comment (pzaj2) [inline src/npc.h]: You're right about the test. I will change the order and use non-equality test then.

File: src/game.cpp (hunk 2 above)
Comment (pzaj2) [inline src/game.cpp]: :+1: will do

Comment (pzaj2) [inline src/game.cpp]: I think you may be right. I will analyze it again and check if this could be removed.

Comment (pzaj2) [inline src/game.cpp]: You're right. The redundant check is removed.

Comment (pzaj2) [inline src/game.cpp]: Seems like this check was redundant at first, removed it.

File: src/npc.h (hunk 3 above)
Comment (pzaj2) [inline src/npc.h]: done

**Comment by ranisalt** [comment]
Thanks!

**Pull request merged**